thorlcr/master/thgraphmanager.cpp

```diff
@@ -1109,31 +1111,89 @@ void thorMain(ILogMsgHandler *logHandler, const char *wuid, const char *graphNam
+                    SocketEndpoint dummyAgentEp;
+                    jobManager->execute(workunit, wuid, currentGraphName, dummyAgentEp);
+                    IException *e = jobManager->queryExitException();
+                    Owned<IWorkUnit> w = &workunit->lock();
+                    if (e)
+                    {
+                        Owned<IWUException> we = w->createException();
+                        we->setSeverity(SeverityInformation);
+                        StringBuffer errStr;
+                        e->errorMessage(errStr);
+                        we->setExceptionMessage(errStr);
+                        we->setExceptionSource("thormasterexception");
+                        we->setExceptionCode(e->errorCode());
+
+                        w->setState(WUStateWait);
+                        break;
+                    }
+
+                    if (lingerPeriod)
+                        w->setDebugValue(instance, "1", true);
+
+                    w->setState(WUStateWait);
                 }
-                w->setState(WUStateWait);
-            }  
-            else
-                jobManager->run();
+                currentGraphName.clear();
+
+                if (lingerPeriod)
+                {
+                    CMessageBuffer msg;
+                    StringBuffer next;
+                    CTimeMon timer(lingerPeriod);
+                    unsigned remaining;
+                    while (!timer.timedout(&remaining))
+                    {
+                        PROGLOG("Lingering time left: %.2f", ((float)remaining)/1000);
+                        if (!queryWorldCommunicator().recv(msg, NULL, MPTAG_THOR, nullptr, remaining))
+                            break;
+                        msg.read(next);
+
+                        // validate
+                        StringArray sArray;
+                        sArray.appendList(next, "/");
+                        if (2 == sArray.ordinality() && streq(sArray.item(0), wuid))
+                        {
+                            currentGraphName.set(sArray.item(1));
+                            // NB: agent could send empty graphName to terminate early before timeout
+                            msg.clear().append(true);
+                            if (!queryWorldCommunicator().reply(msg, 60*1000)) // should be quick!
+                                currentGraphName.clear();
+                            break;
+                        }
+                        // reject/ignore duff message.
+                    }
+
+                    // remove lingering instance from workunit
+                    Owned<IWorkUnitFactory> factory;
+                    Owned<IConstWorkUnit> workunit;
+                    factory.setown(getWorkUnitFactory());
+                    workunit.setown(factory->openWorkUnit(wuid));
+                    Owned<IWorkUnit> w = &workunit->lock();
+                    w->setDebugValue(instance, "0", true);
```

Comment:
Who's responsibility is it to set this value?  I think it should only do this if the workunit has timed out - otherwise the code in executeGraphOnLingeringThor() will have already cleared it, and this will be locking/unlocking the workunit for no benefit.

Reply:
This will also potentially be blocked by the eclagent still holding a IWorkunit, and could also deadlock if there were multiple lingering thors...  Deadlock on IWorkUnit locks need more thought/careful review.

Reply:
Even worse... if this is timing out because eclagent is in a long operation then I think eclagent will still have a read lock on the workunit (need to check carefully), which means this will probably deadlock.

Reply:
> if this is timing out because eclagent is in a long operation then I think eclagent will still have a read lock on the workunit

@ghalliday - does eclagent (or anything) keep a longer term lock on the workunit?

Reply:
If it does, that causes all sorts of problems as soon as it reaches out to anything else which also wants to update the workunit..

Reply:
Looking workunit (openWorkUnit) and the misleading named wuRead.. it is never read locked.
openWorkUnit() does not establish a lock, only a connection.

When eclagent or anything else needs to update it calls lock() and gets an exclusive (write) lock to the workunit.

So that does mean in many places (existing eclagent code and other places), a connection to a workunit could be being read, whilst another thread/process is updating the workunit..
Not great, but not new..

It also means that a reader (without a read lock) doesn't see new changes.. I think that's why there's forceReload() type calls in various places.

I think it is assumed in general, that only 1 thread or 1 process is reading or writing.

This whole area could do with revisiting, but beyond the scope of this PR.

Reply:
So quite a few objections don't actually hold.  I'll look at the PR again to see if that introduces other complications.

Reply:
@jakesmith I think the original comment still stands.  Either eclagent should set this to 0, or thor should (in the non-timeout case), but currently both try to.

Reply:
I had only intended to clear it on timeout, not when it had received next graph, where it had already been cleared.. so yeah I agree, will fix that.